src/cmd/tools/split_shards/main/main.go

```diff
@@ -0,0 +1,337 @@
+	"fmt"
+	"io"
+	iofs "io/fs"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pborman/getopt"
+	"go.uber.org/zap"
+
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/sharding"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+var checkpointPattern = regexp.MustCompile(`/data/(\w+)/([0-9]+)/fileset-([0-9]+)-([0-9]+)-checkpoint.db$`)
+
+func main() {
+	var (
+		optSrcPath       = getopt.StringLong("src-path", 's', "", "Source path [e.g. /temp/lib/m3db/data]")
+		optDstPathPrefix = getopt.StringLong("dst-path", 'd', "", "Destination path prefix [e.g. /var/lib/m3db/data]")
+		optBlockUntil    = getopt.Int64Long("block-until", 'b', 0, "Block Until Time [in nsec]")
+		optShards        = getopt.Uint32Long("src-shards", 'h', 0, "Original (source) number of shards")
+		optFactor        = getopt.IntLong("factor", 'f', 0, "Integer factor to increase the number of shards by")
+	)
+	getopt.Parse()
+
+	rawLogger, err := zap.NewDevelopment()
+	if err != nil {
+		log.Fatalf("unable to create logger: %+v", err)
+	}
+	logger := rawLogger.Sugar()
+
+	if *optSrcPath == "" ||
+		*optDstPathPrefix == "" ||
+		*optSrcPath == *optDstPathPrefix ||
+		*optBlockUntil <= 0 ||
+		*optShards == 0 ||
+		*optFactor <= 0 {
+		getopt.Usage()
+		os.Exit(1)
+	}
+
+	var (
+		srcFilesetLocation = dropDataSuffix(*optSrcPath)
+		dstFilesetLocation = dropDataSuffix(*optDstPathPrefix)
+
+		srcFsOpts = fs.NewOptions().SetFilePathPrefix(srcFilesetLocation)
+		dstFsOpts = fs.NewOptions().SetFilePathPrefix(dstFilesetLocation)
+	)
+
+	// Not using bytes pool with streaming reads/writes to avoid the fixed memory overhead.
+	srcReader, err := fs.NewReader(nil, srcFsOpts)
+	if err != nil {
+		logger.Fatalf("could not create srcReader: %+v", err)
+	}
+
+	dstReaders := make([]fs.DataFileSetReader, *optFactor)
+	dstWriters := make([]fs.StreamingWriter, *optFactor)
+	for i := range dstWriters {
+		dstReaders[i], err = fs.NewReader(nil, dstFsOpts)
+		if err != nil {
+			logger.Fatalf("could not create dstReader, %+v", err)
+		}
+		dstWriters[i], err = fs.NewStreamingWriter(dstFsOpts)
+		if err != nil {
+			logger.Fatalf("could not create writer, %+v", err)
+		}
+	}
+
+	hashFn := sharding.DefaultHashFn(int(*optShards) * (*optFactor))
+
+	start := time.Now()
+
+	if err := filepath.WalkDir(*optSrcPath, func(path string, d iofs.DirEntry, err error) error {
+		if err != nil || d.IsDir() || !strings.HasSuffix(d.Name(), "-checkpoint.db") {
+			return err
+		}
+		fmt.Printf("%s - %s\n", time.Now().Local(), path) // nolint: forbidigo
+		pathParts := checkpointPattern.FindStringSubmatch(path)
+		if len(pathParts) != 5 {
+			return fmt.Errorf("failed to parse path %s", path)
+		}
+
+		var (
+			namespace     = pathParts[1]
+			shard, _      = strconv.Atoi(pathParts[2])
+			blockStart, _ = strconv.Atoi(pathParts[3])
+			volume, _     = strconv.Atoi(pathParts[4])
```

Comment:
Should these be checked for errors returned?

Reply:
These parts are extracted by numeric regexp groups, but perhaps you are right, I'll add the checks just in case.